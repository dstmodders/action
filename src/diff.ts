import * as Diff from 'diff';

interface DiffChange {
  added?: boolean | undefined;
  count: number;
  removed?: boolean | undefined;
  value: string;
}

interface DiffEntry {
  action: string;
  endLine?: number;
  previousValue?: string;
  startLine: number;
  value: string;
}

function newEmptyDiffEntries(): DiffEntry[] {
  const result: DiffEntry[] = [];
  result.pop();
  return result;
}

async function compare(original, changed): Promise<DiffEntry[]> {
  const entries = newEmptyDiffEntries();
  const result: DiffChange[] = Diff.diffLines(original, changed);
  const skip: number[] = [];
  let line: number = 1;

  result.forEach((part, idx) => {
    const nextPart = idx + 1 <= result.length ? result[idx + 1] : null;

    if (!part.removed && !part.added) {
      // unchanged
      line += part.count;
    } else if (part.removed && !part.added && nextPart) {
      const removeLine: number = line;
      line += part.count;

      if (!nextPart.removed && nextPart.added) {
        // replaced
        skip.push(idx + 1);
        entries.push({
          action: 'replace',
          endLine: line,
          previousValue: part.value,
          startLine: removeLine,
          value: nextPart.value,
        });
      } else {
        // removed
        entries.push({
          action: 'remove',
          startLine: line,
          value: part.value,
        });
      }
    } else if (part.added && !skip.includes(idx)) {
      // added
      entries.push({
        action: 'add',
        startLine: line,
        value: part.value,
      });
    } else if (part.removed && !part.added) {
      // removed
      entries.push({
        action: 'remove',
        startLine: line,
        value: part.value,
      });
    }
  });

  return entries;
}

export { DiffChange, DiffEntry, compare, newEmptyDiffEntries };
