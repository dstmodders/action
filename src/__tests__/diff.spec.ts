import expect from 'expect';
import * as diff from '../diff';

describe('diff module', () => {
  describe('compare()', () => {
    const original: string = `1st line
2nd line
3d line
4th line
5th line
6th line
7th line
`;

    describe('when adding', () => {
      const changed: string = `1st line
2nd line
Added line after 2nd line
3d line
Added line after 3d line
4th line
5th line
6th line
7th line
Added line after 7th line
`;

      it('should match the snapshot', async () => {
        expect(await diff.compare(original, changed)).toMatchSnapshot();
      });
    });

    describe('when removing', () => {
      const changed: string = `1st line
4th line
5th line
7th line
`;

      it('should match the snapshot', async () => {
        expect(await diff.compare(original, changed)).toMatchSnapshot();
      });
    });

    describe('when replacing', () => {
      const changed: string = `1st line
Replaced 2nd line
Replaced 3d line
4th line
Replaced 5th line
6th line
Replaced 7th line
`;

      it('should match the snapshot', async () => {
        expect(await diff.compare(original, changed)).toMatchSnapshot();
      });
    });

    describe('when unchanged', () => {
      it('should match the snapshot', async () => {
        expect(await diff.compare(original, original)).toMatchSnapshot();
      });
    });
  });
});
