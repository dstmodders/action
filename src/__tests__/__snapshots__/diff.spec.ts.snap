// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff module compare() when adding should match the snapshot 1`] = `
Array [
  Object {
    "action": "add",
    "startLine": 3,
    "value": "Added line after 2nd line
",
  },
  Object {
    "action": "add",
    "startLine": 4,
    "value": "Added line after 3d line
",
  },
  Object {
    "action": "add",
    "startLine": 8,
    "value": "Added line after 7th line
",
  },
]
`;

exports[`diff module compare() when removing should match the snapshot 1`] = `
Array [
  Object {
    "action": "remove",
    "startLine": 4,
    "value": "2nd line
3d line
",
  },
  Object {
    "action": "remove",
    "startLine": 7,
    "value": "6th line
",
  },
]
`;

exports[`diff module compare() when replacing should match the snapshot 1`] = `
Array [
  Object {
    "action": "replace",
    "endLine": 4,
    "previousValue": "2nd line
3d line
",
    "startLine": 2,
    "value": "Replaced 2nd line
Replaced 3d line
",
  },
  Object {
    "action": "replace",
    "endLine": 6,
    "previousValue": "5th line
",
    "startLine": 5,
    "value": "Replaced 5th line
",
  },
  Object {
    "action": "replace",
    "endLine": 8,
    "previousValue": "7th line
",
    "startLine": 7,
    "value": "Replaced 7th line
",
  },
]
`;

exports[`diff module compare() when unchanged should match the snapshot 1`] = `Array []`;
